### ISSUE TITLE
CRITICAL BUG - infinit query reaches stuck state, unable to call refetch afterwards

### ISSUE THREAD

**Issue by iOSonntag** [D-James-GH/cached_query#32]
There is critical bug inside the InfinitQuery that results into queries that are no longer usable and stuck.

The issue arises if the `_currentFuture` variable fails to be reset to `null` after computations are done. Thus every call to `Query.refetch` will have a none null `_currentFuture` object (which is an already completed but not null Future) and therefore skips the process entirely.

This is very hard to debug and i am still not sure when this issue happens but it has surely to do with race conditions and updating the query from all sides and requesting data from it nearly simoutansously.

But from my point of view the only logical place where this issue can be, is in the `_refetch()` and `_fetchNextPage()` function above the last try catch block, as the _currentFuture gets only reset in the `finally` call of the last try catch.

Everything that happens above the last try catch block might be responsible for the stuck `_currentFuture` as the result from _refetch() gets stored directly into the `_currentFuture` as a Future - my guess would be that if the code aborts above the last try catch, that the final call `_currentFuture = null` is never executed.

So basically if this blocks throws an exception or the return is reached the Query is stuck:
```dart
...
if ((_state.data.isNullOrEmpty || _state.status == QueryStatus.initial) &&
    config.storeQuery) {
  // try to get any data from storage if the query has no data
  final dataFromStorage = await _fetchFromStorage();
  if (dataFromStorage != null) {
    _setState(
      _state.copyWith(
        data: dataFromStorage,
        status: QueryStatus.success,
      ),
    );
    _emit();
    final shouldRefetch = config.shouldRefetch?.call(this, true) ?? true;
    if (!shouldRefetch) {
      return;
    }
  }
}
...
```

Why CRITICAL BUG you ask? Well that ultimatly means the user has to exit the app, shut it down and restart the app in order for him to be able to use pull to refresh again

**Comment by iOSonntag** [comment]
If pull requests are welcome I would try to debug it myself and commit a pull request

**Comment by iOSonntag** [comment]
https://github.com/D-James-GH/cached_query/assets/81633235/1fbc500b-c533-4075-bb88-8951e138f28f

Unfortunatly my PR does not fix the problem I just included the fork as dep override in my project and the issue continues:

<img width="876" alt="image" src="https://github.com/D-James-GH/cached_query/assets/81633235/2544570e-4c8b-4311-8a55-cf4b29713da7">

Drives me crazy :D, if desperitly want to fix that

**Comment by D-James-GH** [comment]
Hey, thank you for bringing this up. 
I believe you are correct in your thinking, if so, moving the try that is already there to the top of the _refetch method should fix it as finally will alway be called then, removing the current future. In the future it will probably be good to have detailed errors rather than a catch all, however, this will go well with your feature request of more detailed loading status. For now having a catch all "try" should fix this issue. I will try to have a look at this tonight, unless you want to have a go on your fork.

**Comment by iOSonntag** [comment]
Using a completer instead of a future fixed the bug. PR #35

**Comment by D-James-GH** [comment]
fix published on cached_query: ^2.0.3
